### PR TITLE
Extension of ovus to:

### PR DESCRIPTION
**Pull request by LeForgeron**
* allow the center of the top sphere to not be on the top of the bottom sphere (via "distance xxx")"
* allow the radius of the torus spindle to be manually specified (via "radius xxx")
* allow precision of the root solver to be tuned ( via "precision xxx")
And change of UV-Mapping to be similar to the uv-mapping of cone & cylinder

### CODE REVIEW

File: source/core/shape/ovus.cpp [hunk 1]
```diff
@@ -45,16 +45,18 @@
...
-*     bottom_radius,top_radius
+*     bottom_radius,top_radius [distance d] [radius r] [precision p]   [sturm]
 *  }
 *
 *  The so long awaited 'Egg' forms.
 *
 *  Normally, the bottom_radius is bigger than the top_radius
-*  the center of the top sphere is at the zenith of the bottom sphere
+*  the center of the top sphere is at the zenith of the bottom sphere, unless
+*    a greater distance d is specified with "distance"
```
Comment (c-lipka) [inline source/core/shape/ovus.cpp]: Is the distance `d` an offset to the default position, or the Y(?) coordinate of the top sphere's center?

Comment (LeForgeron) [inline source/core/shape/ovus.cpp]: It's the distance from the origin, it must be at least the radius of the bottom sphere, but can be bigger (previously, it was always the radius of the bottom sphere)

File: source/core/shape/ovus.cpp [hunk 2]
```diff
@@ -883,46 +896,59 @@ void Ovus::UVCoord(Vector2d& Result, const Intersection *Inter, TraceThreadData
...
-    // now assume it's just a sphere, for UV mapping/projection
-    len = sqrt(x * x + y * y + z * z);
+    // Determine its angle from the point (0, 0, 0) in the x-z plane.
+    len = x * x + z * z;
 
-    if (len == 0.0)
-        return;
+    if ((P[Y]>EPSILON)&&(P[Y]<(VerticalSpherePosition-EPSILON)))
+    {
+    // when not on a face, the range 0.25 to 0.75 is used (just plain magic 25% for face, no other reason, but it makes C-Lipka happy)
```
Comment (c-lipka) [inline source/core/shape/ovus.cpp]: If you really want to make C-Lipka happy, please amend this comment ;)

Dividing at 1/4 and 3/4 has the advantage of the division being exactly at a pixel boundary if the texture is an image 2^N by 2^M pixels in size, which is common for image textures originally designed for mesh-based renderers. It also happens to work for 20*N by 20*M pixels, which is common for image textures with "arbitrary" sizes.

File: source/core/shape/ovus.cpp [hunk 3]
```diff
@@ -883,46 +896,59 @@ void Ovus::UVCoord(Vector2d& Result, const Intersection *Inter, TraceThreadData
...
     z = P[Z];
 
-    // now assume it's just a sphere, for UV mapping/projection
-    len = sqrt(x * x + y * y + z * z);
+    // Determine its angle from the point (0, 0, 0) in the x-z plane.
+    len = x * x + z * z;
 
-    if (len == 0.0)
-        return;
+    if ((P[Y]>EPSILON)&&(P[Y]<(VerticalSpherePosition-EPSILON)))
```
Comment (c-lipka) [inline source/core/shape/ovus.cpp]: Can we get a bit more clarity in this expression (and a few others) by more liberal use of whitespace?

File: source/core/shape/ovus.cpp [hunk 4]
```diff
@@ -820,9 +827,15 @@ void Ovus::Compute_BBox()
...
+    if (!BottomRadius)
+    {
+        bottom = VerticalSpherePosition-TopRadius;
+        length = 2*TopRadius;
+    }
 
-    Make_BBox(BBox, -biggest, -BottomRadius, -biggest,
-              2.0 * biggest, 2.0 * BottomRadius + TopRadius, 2.0 * biggest);
+    Make_BBox(BBox, -biggest, bottom, -biggest,
+            2.0 * biggest, length, 2.0 * biggest);
```
Comment (c-lipka) [inline source/core/shape/ovus.cpp]: Please indent to after opening brace (same as it was before) for clarity.

File: source/core/shape/ovus.cpp [hunk 5]
```diff
@@ -883,46 +896,59 @@ void Ovus::UVCoord(Vector2d& Result, const Intersection *Inter, TraceThreadData
...
-        return;
+    if ((P[Y]>EPSILON)&&(P[Y]<(VerticalSpherePosition-EPSILON)))
+    {
+    // when not on a face, the range 0.25 to 0.75 is used (just plain magic 25% for face, no other reason, but it makes C-Lipka happy)
+        phi = 0.75-0.5*(P[Y])/(VerticalSpherePosition);
+    }
+    else if (P[Y]>EPSILON)
+    {
+    // aka P[Y] is above VerticalSpherePositon, use TopRadius, from 0% to 25%
+       phi = 0.0;
```
Comment (c-lipka) [inline source/core/shape/ovus.cpp]: Please stick to a 4-blanks indent.
Also, please indent the comment inside the braces.

File: source/core/shape/ovus.cpp [hunk 6]
```diff
@@ -883,46 +896,59 @@ void Ovus::UVCoord(Vector2d& Result, const Intersection *Inter, TraceThreadData
...
+            phi = (sin(sqrt(1-t)*M_PI_2)/(4.0));
+       }
+    }
     else
     {
-        x /= len;
-        y /= len;
-        z /= len;
+    // aka P[Y] is below origin (<0), use BottomRadius, from 75% to 100%
+       phi = 1.0;
```
Comment (c-lipka) [inline source/core/shape/ovus.cpp]: Indent, see above

File: source/core/shape/ovus.cpp [hunk 7]
```diff
@@ -883,46 +896,59 @@ void Ovus::UVCoord(Vector2d& Result, const Intersection *Inter, TraceThreadData
...
+       if (BottomRadius)
+       {
+          t = ((BottomRadius+P[Y])/(BottomRadius));
```
Comment (c-lipka) [inline source/core/shape/ovus.cpp]: Indent, see above

File: source/core/shape/ovus.cpp [hunk 8]
```diff
@@ -883,46 +896,59 @@ void Ovus::UVCoord(Vector2d& Result, const Intersection *Inter, TraceThreadData
...
+          phi = 1.0-sin(sqrt(t)*M_PI_2)/(4.0);
+       }
+       else if (TopRadius)
+       {
+        // degenerate ovus in sphere
```
Comment (c-lipka) [inline source/core/shape/ovus.cpp]: Indent, see above

File: source/parser/parser.cpp [hunk 9]
```diff
@@ -4892,51 +4894,115 @@ ObjectPtr Parser::Parse_Ovus()
...
-            Object->TopVertical = ((Object->BottomRadius - Object->VerticalPosition) * Object->TopRadius / distance ) + Object->BottomRadius;
-        } else {
-            Object->ConnectingRadius = 2.0 * Object->TopRadius;
-            Object->VerticalPosition = - 2.0 * Object->TopRadius + Object->BottomRadius + (1.5 * Object->TopRadius * Object->TopRadius / Object->BottomRadius);
-            Object->HorizontalPosition = sqrt((Object->TopRadius)*(Object->TopRadius) - ((Object->VerticalPosition - Object->BottomRadius) * (Object->VerticalPosition - Object->BottomRadius)));
-            Object->TopVertical = 2.0 * Object->BottomRadius - Object->VerticalPosition;
-            distance = Object->ConnectingRadius - Object->BottomRadius;
-            Object->BottomVertical = -Object->VerticalPosition * Object->BottomRadius / distance;
-        }
+        Error("distance of Ovus must be greater than bottom radius");
```
Comment (c-lipka) [inline source/parser/parser.cpp]: Why? Shouldn't the ovus object also work with a lower VSP, provided VSP+TR > BR?

**Comment by c-lipka** [comment]
From what I see here, when you created the pull request you seem to have checked a box granting reviewers (such as myself) write access to the branch corresponding to the pull request (`feature-extended-ovus` in your fork of the GitHub repo), allowing them to directly modify your pull request; let me know if this was intentional or otherwise.

Also note that any further commits you push to the `feature-extended-ovus` branch will automatically be picked up as amendments to this pull request.

**Comment by LeForgeron** [comment]
> From what I see here, when you created the pull request you seem to have checked a box granting reviewers (such as myself) write access to the branch corresponding to the pull request (feature-extended-ovus in your fork of the GitHub repo), allowing them to directly modify your pull request; let me know if this was intentional or otherwise.

Not intentional, probably default, but welcomed.

> Also note that any further commits you push to the feature-extended-ovus branch will automatically be picked up as amendments to this pull request.

Ok, might work on it tomorrow, if I still need to.

**Pull request merged**